{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removes CPU usage limit by some jupyter versions\n",
    "import os\n",
    "os.environ['KMP_AFFINITY'] = ''\n",
    "# Configure matplotlib to enable large animations\n",
    "import matplotlib\n",
    "matplotlib.rcParams['animation.embed_limit'] = 2**128\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import dictys\n",
    "\n",
    "# paths\n",
    "dictys_data_path = \"/ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/tut_files/skin/data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dist DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S0</th>\n",
       "      <th>S1</th>\n",
       "      <th>S2</th>\n",
       "      <th>S3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Trial60.skin.R1.01.R2.02.R3.21.P1.56</th>\n",
       "      <td>1.796973</td>\n",
       "      <td>2.377683</td>\n",
       "      <td>10.384121</td>\n",
       "      <td>4.781205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trial60.skin.R1.01.R2.03.R3.61.P1.54</th>\n",
       "      <td>8.058921</td>\n",
       "      <td>12.233577</td>\n",
       "      <td>0.528227</td>\n",
       "      <td>11.043153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trial60.skin.R1.01.R2.06.R3.64.P1.56</th>\n",
       "      <td>3.439011</td>\n",
       "      <td>0.735645</td>\n",
       "      <td>12.026159</td>\n",
       "      <td>6.423243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trial60.skin.R1.01.R2.06.R3.83.P1.55</th>\n",
       "      <td>1.028344</td>\n",
       "      <td>3.146313</td>\n",
       "      <td>9.615491</td>\n",
       "      <td>4.012575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trial60.skin.R1.01.R2.11.R3.86.P1.56</th>\n",
       "      <td>1.677695</td>\n",
       "      <td>2.496962</td>\n",
       "      <td>10.264842</td>\n",
       "      <td>4.661926</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            S0         S1         S2  \\\n",
       "Trial60.skin.R1.01.R2.02.R3.21.P1.56  1.796973   2.377683  10.384121   \n",
       "Trial60.skin.R1.01.R2.03.R3.61.P1.54  8.058921  12.233577   0.528227   \n",
       "Trial60.skin.R1.01.R2.06.R3.64.P1.56  3.439011   0.735645  12.026159   \n",
       "Trial60.skin.R1.01.R2.06.R3.83.P1.55  1.028344   3.146313   9.615491   \n",
       "Trial60.skin.R1.01.R2.11.R3.86.P1.56  1.677695   2.496962  10.264842   \n",
       "\n",
       "                                             S3  \n",
       "Trial60.skin.R1.01.R2.02.R3.21.P1.56   4.781205  \n",
       "Trial60.skin.R1.01.R2.03.R3.61.P1.54  11.043153  \n",
       "Trial60.skin.R1.01.R2.06.R3.64.P1.56   6.423243  \n",
       "Trial60.skin.R1.01.R2.06.R3.83.P1.55   4.012575  \n",
       "Trial60.skin.R1.01.R2.11.R3.86.P1.56   4.661926  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dist DataFrame shape: (6436, 4)\n",
      "edge DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1\n",
       "0  0  1\n",
       "1  0  2\n",
       "2  0  3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "branch DataFrame:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0  1\n",
       "0  0  1\n",
       "1  0  2\n",
       "2  0  1\n",
       "3  0  1\n",
       "4  0  1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "branch DataFrame shape: (6436, 2)\n"
     ]
    }
   ],
   "source": [
    "# Load data\n",
    "dist = pd.read_csv(os.path.join(dictys_data_path, 'tmp/dist.tsv.gz'), header=0, index_col=0, sep='\\t')\n",
    "edge = pd.read_csv(os.path.join(dictys_data_path, 'tmp/edge.tsv.gz'), header=None, index_col=None, sep='\\t')\n",
    "branch = pd.read_csv(os.path.join(dictys_data_path, 'tmp/branch.tsv.gz'), header=None, index_col=None, sep='\\t')\n",
    "\n",
    "# Display the first few rows of each DataFrame to confirm successful loading\n",
    "print(\"dist DataFrame:\")\n",
    "display(dist.head())\n",
    "print(\"dist DataFrame shape:\", dist.shape)\n",
    "\n",
    "print(\"edge DataFrame:\")\n",
    "display(edge.head())\n",
    "\n",
    "print(\"branch DataFrame:\")\n",
    "display(branch.head())\n",
    "print(\"branch DataFrame shape:\", branch.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/ocean/projects/cis240075p/asachan/.conda/envs/scvelo/lib/python3.8/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import os \n",
    "\n",
    "# load adata\n",
    "adata = sc.read_h5ad(os.path.join('/ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/tut_files/skin/data', 'tmp/adata_with_traj.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12.761804034895938\n",
      "12.761804034895938\n",
      "11.571379671158407\n",
      "8.587147894050306\n"
     ]
    }
   ],
   "source": [
    "# get max pseudotime for each branch\n",
    "S1_max_pseudotime = adata.obs['S1_pseudotime'].max()\n",
    "S2_max_pseudotime = adata.obs['S2_pseudotime'].max()\n",
    "S3_max_pseudotime = adata.obs['S3_pseudotime'].max()\n",
    "S0_max_pseudotime = adata.obs['S0_pseudotime'].max()\n",
    "print(S1_max_pseudotime)\n",
    "print(S2_max_pseudotime)\n",
    "print(S3_max_pseudotime)\n",
    "print(S0_max_pseudotime)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "traj = dictys.traj.trajectory.fromdist(edge.values, dist.values)\n",
    "traj_file_path = os.path.join(dictys_data_path, 'traj_node.h5')\n",
    "traj.to_file(traj_file_path)\n",
    "\n",
    "point = dictys.traj.point.fromdist(traj, branch.values, dist.values)\n",
    "point_file_path = os.path.join(dictys_data_path, 'traj_cell_rna.h5')\n",
    "point.to_file(point_file_path, traj=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare genome related files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading reference genome mm10 in homer\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/tut_files/skin/data\n",
    "\n",
    "# Motifs (file motifs.motif)\n",
    "# Option 1: from HOCOMOCO (https://hocomoco11.autosome.org/)\n",
    "wget -q -o /dev/null -O - 'https://hocomoco11.autosome.org/final_bundle/hocomoco11/full/MOUSE/mono/HOCOMOCOv11_full_MOUSE_mono_homer_format_0.0001.motif' | awk -F \"\\t\" 'BEGIN { OFS = \"\\t\"} {if (substr($1,1,1) == \">\") $2=substr($2,1,1)tolower(substr($2,2)); print}' > motifs.motif\n",
    "# Option 2: from HOMER\n",
    "# dictys_helper motif_homer.sh -c 3 > motifs.motif\n",
    "# Option 3: provide your custom motifs\n",
    "\n",
    "# Reference genome (folder genome)\n",
    "# Note: You need the same reference genome version with chromatin accessibility reads\n",
    "# Option 1: download genome from HOMER\n",
    "dictys_helper genome_homer.sh mm10 genome\n",
    "# Option 2: provide your custom genome\n",
    "\n",
    "# Bed file for TSS (file gene.bed)\n",
    "# Download gtf file from ensembl\n",
    "wget -q -o /dev/null -O gene.gtf.gz http://ftp.ensembl.org/pub/release-102/gtf/mus_musculus/Mus_musculus.GRCm38.102.gtf.gz\n",
    "gunzip gene.gtf.gz\n",
    "# Convert to bed\n",
    "dictys_helper gene_gtf.sh gene.gtf gene.bed\n",
    "rm gene.gtf\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dictys",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
